pkg/app/piped/planner/kubernetes/kubernetes.go

```diff
@@ -421,17 +421,29 @@ func checkImageChange(ns diff.Nodes) (string, bool) {
 	return desc, true
 }
 
+// checkReplicasChange checks if the replicas field is changed.
 func checkReplicasChange(ns diff.Nodes) (before, after string, changed bool) {
 	const replicasQuery = `^spec\.replicas$`
 	node, err := ns.FindOne(replicasQuery)
 	if err != nil {
-		return
+		// not found or unknown error
+		return "", "", false
 	}
 
 	before = node.StringX()
+	if node.TypeX == nil {
+		// The replicas field is not found in the old manifest.
+		before = ""
+	}
```

Comment:
🤔 oops, we should put this before as `"nil"`?

Reply:
@khanhtc1202 @t-kikuc @ffjlabo 
WDYT about this?

Reply:
@Warashi 
When does `node.TypeX == nil` occur?
I think we should clarify the situation to determine it.

Reply:
@ffjlabo Sorry for forgetting to write it.
When the replicas field is not found in the old manifests, `node.TypeX` becomes `nil` and `node.ValueX` becomes invalid value.

Reply:
@Warashi 

> @ffjlabo Sorry for forgetting to write it.
When the replicas field is not found in the old manifests, node.TypeX becomes nil and node.ValueX becomes invalid value.

Thanks.

> When the replicas field is not found in the old manifests,

It seems that the `node` is the one that has `spec.replicas` extracting by the logic before.
So we might not need to check it. 

```
	const replicasQuery = `^spec\.replicas$`
	node, err := ns.FindOne(replicasQuery)
```

I'm sorry if I misunderstood. 🙏

Reply:
Hmm, I think one side of diff defines spec.replicas, and the other side does not; this occurs.

Reply:
@ffjlabo 
Without this check, this test case fails.
https://github.com/pipe-cd/pipecd/pull/5341/files#diff-4c74652886416486eb7b42cc641e5fd09672015dbaeeb19e3dddefaf81e913dbR1351-R1381

Reply:
@Warashi 
> Hmm, I think one side of diff defines spec.replicas, and the other side does not; this occurs.

I got your point. You are right. Sorry, I misunderstood. 🙏 
We need it in such a case.

IMO, It might be better to init `before` as empty first. I felt easy to guess the value.

```
before = ""
if node.TypeX != nil {
    // found spec.replicas in the old manifests
    before = node.StringX()
}
```

Reply:
I revised the whole, how about this code?

```go
// checkReplicasChange checks if the replicas field is changed.
func checkReplicasChange(ns diff.Nodes) (before, after string, changed bool) {
	const replicasQuery = `^spec\.replicas$`
	node, err := ns.FindOne(replicasQuery)
	if err != nil {
		// no difference between the before and after manifests, or unknown error occurred.
		return "", "", false
	}

	if node.TypeX == nil {
		// The replicas field is not found in the before manifest.
		// There is difference between the before and after manifests, So it means the replicas field is added in the after manifest.
		return "", node.StringY(), true
	}

	if node.TypeY == nil {
		// The replicas field is not found in the after manifest.
		// There is difference between the before and after manifests, So it means the replicas field is removed in the after manifest.
		return node.StringX(), "", true
	}

	return node.StringX(), node.StringY(), true
}
```

Reply:
@Warashi It's also ok 👍 I can understand the code because of the comments.

Reply:
Thanks 👍🏻 

BTW, I realized it's better to return `"<nil>"` instead of an empty string because the before and after are used in this context.
https://github.com/pipe-cd/pipecd/blob/23eddac42ab78523e522ac534437b8718f200ba1/pkg/app/piped/planner/kubernetes/kubernetes.go#L311-L313

Reply:
fixed on this commit
https://github.com/pipe-cd/pipecd/pull/5341/commits/13625cacf2d14b5ff0d3ab0c9701503e8f6dfd50

Reply:
@Warashi
Nice catch! I agree with it!  The case is when there is nothing `spec.replicas`.

I think the `<nil>` value should be defined as const value. WDYT?
We might use it to express the same situation. e.g. checkImageTags.

Reply:
@ffjlabo 
I think it's too early to make `"<nil>"` as constant because we use this only two places in one function, and we should use `"<nil>"` carefully because whether the `"<nil>"` or an empty string should be used is case-by-case.

For example, I think the image `nginx` should be just `nginx`, not `nginx:<nil>`.

Reply:
@Warashi 
I agree with it. Thanks. So let's rethink it when the same situation will occur. 👌